avogadro/core/molecule.cpp

```diff
@@ -737,5 +847,203 @@ Residue& Molecule::residue(Index index)
   return m_residues[index];
 }
 
+bool Molecule::setBondPairs(const Array<std::pair<Index, Index>>& pairs)
+{
+  if (pairs.size() == bondCount()) {
+    Index bond = 0;
+    for (const auto& pair : pairs) {
+      setBondPair(bond, pair);
+      ++bond;
+    }
+    return true;
+  }
+  return false;
+}
+
+bool Molecule::setBondPair(Index bondId, const std::pair<Index, Index>& pair)
+{
+  if (bondId < bondCount()) {
+    if (!m_graphDirty) {
+      m_graph.removeEdge(m_bondPairs[bondId].first, m_bondPairs[bondId].second);
+      m_graph.addEdge(pair.first, pair.second);
+    }
+    m_bondPairs[bondId] = pair;
+    return true;
+  }
+  return false;
+}
+
+unsigned char Molecule::bondOrder(Index bondId) const
+{
+  return bondId < m_bondOrders.size() ? m_bondOrders[bondId] : 0;
+}
+
+bool Molecule::setBondOrders(const Array<unsigned char>& orders)
+{
+  if (orders.size() == bondCount()) {
+    m_bondOrders = orders;
+    return true;
+  }
+  return false;
+}
+
+bool Molecule::setBondOrder(Index bondId, unsigned char order)
+{
+  if (bondId < bondCount()) {
+    m_bondOrders[bondId] = order;
+    return true;
+  }
+  return false;
+}
+
+Index Molecule::atomCount(unsigned char number) const
+{
+  Index count(0);
+  for (Array<unsigned char>::const_iterator it = m_atomicNumbers.begin();
+       it != m_atomicNumbers.end(); ++it) {
+    if (*it == number)
+      ++count;
+  }
+  return count;
+}
+
+bool Molecule::setAtomicNumbers(const Core::Array<unsigned char>& nums)
+{
+  if (nums.size() == atomCount()) {
+    m_atomicNumbers = nums;
+    return true;
+  }
+  return false;
+}
+
+bool Molecule::setAtomicNumber(Index atomId, unsigned char number)
+{
+  if (atomId < atomCount()) {
+    m_atomicNumbers[atomId] = number;
+    return true;
+  }
+  return false;
+}
+
+bool Molecule::hasCustomElements() const
+{
+  for (Array<unsigned char>::const_iterator it = m_atomicNumbers.begin(),
+                                            itEnd = m_atomicNumbers.end();
+       it != itEnd; ++it) {
+    if (Core::isCustomElement(*it))
+      return true;
+  }
+  return false;
+}
+
+std::map<unsigned char, size_t> Molecule::composition() const
+{
+  // A map of atomic symbols to their quantity.
+  std::map<unsigned char, size_t> composition;
+  for (Array<unsigned char>::const_iterator it = m_atomicNumbers.begin(),
+                                            itEnd = m_atomicNumbers.end();
+       it != itEnd; ++it) {
+    ++composition[*it];
+  }
+  return composition;
+}
+
+Index Molecule::findBond(Index atomId1, Index atomId2) const
```

Comment:
I think we can omit "find" and just call it Molecule::bond(Index, index) - it would match Molecule::atom(index) which already exists.

Reply:
There is already a `BondType bond(Index atomId1, Index atomId2) const` I cannot overload the `bond` function. Any alternative name?

Reply:
The same but returns the index. I just realize I can use `bond` and extract the index.